src/binary-encoding.adoc

```diff
@@ -2023,3 +2022,292 @@ ED 05 93 C6
       └─┬─┘
 NOP bytes, values ignored
 ----
+
+[[e_expression_arguments]]
+=== E-expression Arguments
+
+The binary encoding of E-expressions (aka macro invocations) starts with the address of the macro to expand. The address
+can be encoded as <<e_expression_with_the_address_in_the_opcode, part of the opcode>> or as
+<<e_expression_with_the_address_as_a_trailing_flexuint, a `FlexUInt` that follows the opcode>>.
+
+The encoding of the E-expression's arguments depends on their respective types. Argument types can be classified as
+belonging to one of two categories: <<tagged_encodings, _tagged encodings_>> and
+<<tagless_encodings, _tagless encodings_>>.
+
+[[tagged_encodings]]
+==== Tagged Encodings
+
+_Tagged types_ are argument types whose encoding begins with an <<opcodes, opcode>>, sometimes informally called a 'tag'.
+These include the <<core_types, core types>> and the <<abstract_types, abstract types>>.
+
+[[core_types]]
+===== Core types
+
+The _core types_ are the 13 types in the Ion data model:
+
+`null`
+| `bool`
+| `int`
+| `float`
+| `decimal`
+| `timestamp`
+| `string`
+| `symbol`
+| `blob`
+| `clob`
+| `list`
+| `sexp`
+| `struct`
+
+[[abstract_types]]
+===== Abstract types
+
+The _abstract types_ are unions of two or more of the <<core_types, core types>>.
+
+[cols="^.^1a, 6a"]
+|===
+|Abstract type |Included Ion types
+
+|`any`
+|All core Ion types
+
+|`number`
+|`int`, `float`, `decimal`
+
+|`exact`
+|`int`, `decimal`
+
+|`text`
+|`string`, `symbol`
+
+|`lob`
+|`blob`, `clob`
+
+|`sequence`
+|`list`, `sexp`
+|===
+
+===== Tagged E-expression Argument Encoding
+
+When a macro parameter has a tagged type, the encoding of that parameter's corresponding argument in an E-expression
+is identical to how it would be encoded anywhere else in an Ion stream: it has a leading <<opcodes, opcode>> that
+dictates how many bytes follow and how they should be interpreted. This is very flexible, but makes it possible
+for writers to encode values that conflict with the parameter's declared type. Because of this, the macro expander will
+read the argument and then check its type against the parameter's declared type. If it does not match, the macro
+expander must raise an error.
+
+Macro `foo` (defined below) is used in this section's subsequent examples to demonstrate the encoding of tagged-type
+arguments.
+
+.Figure {counter:figure}: Definition of example macro `_foo_` at address 0
+[%unbreakable,source]
+----
+(macro
+    foo           // Macro name
+    [(x number!)] // Parameters
+    /*...*/       // Template (elided)
+)
+----
+
+.Figure {counter:figure}: Encoding of E-expression `_(:foo 3.14e)_`
```

Comment:
I suggest using global-address references like `:0` in these examples, it eliminates a point of indirection for readers, and focuses on the parts most relevant to the binary.